private/model/api/passes.go

```diff
@@ -160,36 +159,58 @@ func (a *API) fixStutterNames() {
 	}
 }
 
-func (a *API) validateShapeNames() {
+// validateShapeNames is valid only for shapes of type structure or enums
+// We validate a shape name to check if its a valid shape name
+// A valid shape name would only contain alphanumeric characters and have an alphabet as leading character.
+//
+// If we encounter a shape name with underscores(_), we remove the underscores, and
+// follow a canonical upper camel case naming scheme to create a new shape name.
+// If the shape name collides with an existing shape name we return an error.
+// The resulting shape name must be a valid shape name or throw an error.
+func (a *API) validateShapeNames() error {
 	for _, s := range a.Shapes {
-		// name should start with an alphabet, that is ascii character 65 to 90, and 97 to 122;
-		// else skip that character
-		name := s.ShapeName
-		for len(name) > 1 {
-			if (s.Type == "structure" || s.IsEnum()) && !unicode.IsLetter(rune(name[0])) {
-				// Remove the leading underscores from the name of the shape, if shape is enum or structure
-				if name[0] == '_' {
-					name = name[1:]
-				} else {
-					// Throw an error if shape name starts with non alphabetic character and
-					// above condition is unsatisfied.
-					log.Fatalf("Shape starting with non alphabetical character found: %v", s.ShapeName)
-				}
-			} else {
-				break
+		if s.Type == "structure" || s.IsEnum() {
+			name := s.ShapeName
+
+			// The regexp used allows underscores(_) at the beginning of the shape name
+			// There may be 0 or more underscores(_). The next character which would be the leading character
+			// in the renamed shape must be an alphabetic character.
+			// The regex allows alphanumeric characters along with underscores(_) in rest of the string.
+			compiledRegExp, err := regexp.Compile("^[_]*[a-zA-Z][a-zA-Z0-9_]*$")
+			if err != nil {
+				return fmt.Errorf("failed to compile regex while validating shape name")
+			}
+
+			// check if the shape name satisfies the regex
+			if b := compiledRegExp.Match([]byte(name)); !b {
+				return fmt.Errorf("invalid shape name found: %v", s.ShapeName)
 			}
-		}
 
-		if s.ShapeName != name {
-			debugLogger.Logf("Renamed shape %v to %v for package %v \n", s.ShapeName, name, a.PackageName())
-			if a.Shapes[name] != nil {
-				// throw an error if shape with a new shape name already exists
-				log.Fatalf("Tried to rename shape %v to %v, but the new name results in shape name collision",
-					s.ShapeName, name)
+			// replace all _ with whitespace
+			name = strings.Replace(name, "_", " ", -1)
+			// remove whitespace and get a slice of strings
+			slice := strings.Fields(name)
```

Comment:
Could combine the `Replace` and `Fields` statement together with [strings.FieldsFunc](https://golang.org/pkg/strings/#FieldsFunc). This utility will swash multiple delimiters together when separating fields.

https://play.golang.org/p/9NDH5VonVp9